torch/fx/experimental/guard_env.py

```diff
@@ -0,0 +1,102 @@
+from typing import List, Dict
+import torch
+import dataclasses
+from contextlib import contextmanager
+
+__all__ = [
+    "GuardEnvExpr", "DuplicateInputs", "GuardEnv",
+    "CURRENT_GUARD_ENV", "guarding",
+]
+
+
+"""
+Parent structure for guard env expressions.
+
+A GuardEnvExpr can have any subtype.
+
+Note: All subtypes must be handled exhaustively in
+torch._dynamo.guards._parse_guard_env_guards to avoid a RuntimeError.
+"""
+@dataclasses.dataclass
+class GuardEnvExpr():
+    pass
+
+"""
+A class representing a pair of duplicate inputs.
+
+arg_a and arg_b are argument names in the traced scope.
+"""
+@dataclasses.dataclass
+class DuplicateInputs(GuardEnvExpr):
+    arg_a: str
+    arg_b: str
```

Comment:
I think, directionally, my biggest discomfort with introducing a new GuardEnvExpr concept here that is distinct from Dynamo guards, is that Dynamo may very well WANT to generate guards like this (e.g., it is processing a conditional `a is b`) and it seems weird to have two possible ways to generate a guard (regular Guard, or GuardEnvExpr)

Reply:
Agreed, it would be good to consolidate how we manage guards into a single place.  That doesn't mean all guards need to be sympy guards (or dynamo guards)

Reply:
This is due to a layering issue - aot_autograd is below dynamo, so this needs to be in another place. We could violate layering and reuse the definitions in guards.py, or we could lower those down to a shared place, or we could do something like this. This felt the more reasonable, especially given the fact that we have another guard system in shape_env already that follows a similar pattern.

Reply:
This is also different from dynamo guards as all dynamo guards handle a single object at a time, the entire guards static_fn pattern is predicated on passing a Guard object which represent a single arg (see the `self.get` calls in guards.py). We could break that assumption and make multi-target guards, and rewrite that file, or we could carve out a special one-off guard that takes multiple tensors.

Reply:
Fwiw, I understand the concern, and we could do something like:

1) throw out DuplicateInputs
2) Make the mapping tensor:source instead of tensor:name
3) define a new Guard type that is multi-source / or a Source type that is multi-guard (we need to model the fact that this is a 2 object guard somewhere)
4) Call Guards.DUPLICATE_INPUTS with either the 2 object guard, or with 2 guards for the 2 sources

Would you prefer that? Does that feel cleaner to you both? it does not to me.